### DISCUSSION TITLE
Starting and stopping Oxigraph server in the background

### DISCUSSION THREAD

**Discussion by Abbe98** [oxigraph/oxigraph#491] [Ideas]
I'm currently experimenting with using Oxigraph to build [Snowman](https://github.com/glaciers-in-archives/snowman) sites in CI-environments without the need for remote SPARQL endpoints.

It works really well and looks something like:

1. I end up starting the server in the background to ensure it does not block the foreground process.
2. Then I let the foreground sleep for 1 second to ensure the server will start before Snowman starts sending queries.
3. Finally I let the tear-down of the CI take care of stopping the server.

```bash
oxigraph serve --location datastore & sleep 1 && snowman build --timeit
```

The above have two core issues; `oxigraph serve` could fail in the background, one has to manually kill the correct Oxigraph process. **Both these issues could be mitigated with scripting or by bringing in a service manager**, however, it would would be neat if Oxigraph server had build-in support for starting and stopping Oxigraph in the background, given control-flow use cases like this one.

I image it could look something like:

```bash
oxigraph serve start --location datastore && snowman build --timeit && oxigraph serve stop
```

**Comment by adamreichold** [comment]
I think this is integrating policy into how the service is managed into the service itself which is generally avoided. Doesn't something like `supervisord` suffice? Or if this is about a CI system, those usually have integrated mechanisms for providing background services for the CI jobs themselves.

**Comment by Tpt** [comment]
Thank you for using Oxigraph!

That's a great question. I agree with @adamreichold, I am a bit afraid that implementing `oxigraph serve start` and `oxigraph serve stop` will be quite dundant with a a proper service manager. We need to do the same things, spawing a process, checking its health, stopping it, managing logs... What do you think about providing some ready to use configuration files for some service managers instead? But I would be happy to be convinced otherwise, I am mostly trying here to avoid "feature creep". It's already a quite large project.

**Reply by Abbe98** [reply]
I have no strong opinions, my intent here is as much about having the use case noted as it is about making things easier. The provided docker container goes a long way with some CI servers but for more "raw" environments a `systemd` config would be a nice addition. Not sure about other service managers.

**Reply by Tpt** [reply]
Thank you! Yes, a systemd config file would be amazing! It might also be nice to implement support of the systemd "notify" startup system into Oxigraph to allow Systemd to cleanly wait for Oxigraph to be up: https://www.freedesktop.org/software/systemd/man/sd_notify.html#

Bug: #499

**Reply by Abbe98** [reply]
Oh indeed, the notify support does a very nice job of taking care of the issue of needing to wait for Oxigraph to start. Thank you!
